aten/src/ATen/native/xnnpack/Pooling.cpp

```diff
@@ -0,0 +1,321 @@
+          ceil_mode,
+          output_min,
+          output_max),
+      "mobile::cpu::max_pool not available!");
+
+  xnn_operator_t max_pool_op{};
+
+  const xnn_status create_status = xnn_create_max_pooling2d_nhwc_f32(
+      padding[Layout::Parameter::height],   // input_padding_top
+      padding[Layout::Parameter::width],    // input_padding_right
+      padding[Layout::Parameter::height],   // input_padding_bottom
+      padding[Layout::Parameter::width],    // input_padding_left
+      kernel[Layout::Parameter::height],    // kernel_height
+      kernel[Layout::Parameter::width],     // kernel_width
+      stride[Layout::Parameter::height],    // subsampling_height
+      stride[Layout::Parameter::width],     // subsampling_width
+      dilation[Layout::Parameter::height],  // dilation_height
+      dilation[Layout::Parameter::width],   // dilation_width
+      input_channels,                       // channels,
+      input_channels,                       // input_pixel_stride - Always converted to NHWC contiguous prior to use.
+      output_channels,                      // output_pixel_stride
+      output_min,                           // output_min
+      output_max,                           // output_max
+      0u,                                   // flags
+      &max_pool_op);                        // operator
+
+  TORCH_CHECK(
+      xnn_status_success == create_status,
+      "xnn_create_max_pooling2d_nhwc_f32 failed!");
+
+  return Context{
+    Operator(max_pool_op),
+    {
+      output_channels,
+      {
+        kernel[Layout::Parameter::height],
+        padding[Layout::Parameter::height],
+        stride[Layout::Parameter::height],
+        dilation[Layout::Parameter::height],
+        ceil_mode,
+      },
+      {
+        kernel[Layout::Parameter::width],
+        padding[Layout::Parameter::width],
+        stride[Layout::Parameter::width],
+        dilation[Layout::Parameter::width],
+        ceil_mode,
+      },
+    }
+  };
+}
+
+bool shape(const int64_t input, const Context::Output::Spatial& spatial) {
+  return pooling_output_shape(
+      input,
+      spatial.kernel,
+      spatial.padding,
+      spatial.stride,
+      spatial.dilation,
+      spatial.ceil_mode);
+};
+
+bool usable(
+    const Tensor& input,
+    const Context::Output::Spatial& height,
+    const Context::Output::Spatial& width) {
+  using namespace internal;
+
+         // Input
+  return (4 == input.dim()) &&
+         (c10::DeviceType::CPU == input.device().type()) &&
+         (kFloat == input.scalar_type()) &&
+         // Output
+         (shape(input.size(Layout::Activation4D::height), height) > 0) &&
+         (shape(input.size(Layout::Activation4D::width), width) > 0) &&
+         true;
+}
+
+Tensor run(
+    const Context& context,
+    const Tensor& input) {
+  using namespace internal;
+
+  TORCH_CHECK(
+      usable(input, context.output.height, context.output.width),
+      "mobile::cpu::max_pool not usable!");
```

Comment:
Message needs to be updated.  Also, shouldn't this be checked when the context is created?

Reply:
Updated the message.

The input tensor is not available at creation time.  available() checks any parameter provided at creation time while usable gates anything that depends on the input tensor.